install_test/install_test.cpp

```diff
@@ -0,0 +1,312 @@
+    }
+
+    // core/matrix/identity.hpp
+    {
+        using Mtx = gko::matrix::Identity<>;
+        auto test = Mtx::create(refExec);
+    }
+
+    // core/matrix/sellp.hpp
+    {
+        using Mtx = gko::matrix::Sellp<>;
+        auto test = Mtx::create(refExec, gko::dim<2>{2, 2}, 2);
+    }
+
+    // core/preconditioner/jacobi.hpp
+    {
+        using Bj = gko::preconditioner::Jacobi<>;
+        auto test = Bj::build().with_max_block_size(1u).on(refExec);
+    }
+
+    // core/solver/bicgstab.hpp
+    {
+        using Solver = gko::solver::Bicgstab<>;
+        auto test = Solver::build()
+                        .with_criteria(
+                            gko::stop::Iteration::build().with_max_iters(1u).on(
+                                refExec))
+                        .on(refExec);
+    }
+
+    // core/solver/cg.hpp
+    {
+        using Solver = gko::solver::Cg<>;
+        auto test = Solver::build()
+                        .with_criteria(
+                            gko::stop::Iteration::build().with_max_iters(1u).on(
+                                refExec))
+                        .on(refExec);
+    }
+
+    // core/solver/cgs.hpp
+    {
+        using Solver = gko::solver::Cgs<>;
+        auto test = Solver::build()
+                        .with_criteria(
+                            gko::stop::Iteration::build().with_max_iters(1u).on(
+                                refExec))
+                        .on(refExec);
+    }
+
+    // core/solver/fcg.hpp
+    {
+        using Solver = gko::solver::Fcg<>;
+        auto test = Solver::build()
+                        .with_criteria(
+                            gko::stop::Iteration::build().with_max_iters(1u).on(
+                                refExec))
+                        .on(refExec);
+    }
+
+    // core/solver/gmres.hpp
+    {
+        using Solver = gko::solver::Gmres<>;
+        auto test = Solver::build()
+                        .with_criteria(
+                            gko::stop::Iteration::build().with_max_iters(1u).on(
+                                refExec))
+                        .on(refExec);
+    }
+
+    // core/stop/
+    {
+        // iteration.hpp
+        auto iteration =
+            gko::stop::Iteration::build().with_max_iters(1u).on(refExec);
+
+        // time.hpp
+        auto time = gko::stop::Time::build()
+                        .with_time_limit(std::chrono::milliseconds(10))
+                        .on(refExec);
+        // residual_norm_reduction.hpp
+        auto res_red = gko::stop::ResidualNormReduction<>::build()
+                           .with_reduction_factor(1e-10)
+                           .on(refExec);
+
+        // stopping_status.hpp
+        auto stop_status = gko::stopping_status{};
+
+        // combined.hpp
+        auto combined =
+            gko::stop::Combined::build()
+                .with_criteria(std::move(time), std::move(iteration))
+                .on(refExec);
+    }
+
+
+    return 0;
```

Comment:
Nit: never more than one empty line inside a function, as per [whitespace guidelines](/home/gflegar/Sources/ginkgo/test/install). And you don't need a return statement, as it's inserted automatically ([special property \#4](https://en.cppreference.com/w/cpp/language/main_function)).

Reply:
I left the return statement because at some point I got a warning that I do not have a return statement in my main (I did not set it void because I like to have a return statement and explicitly state the error code the program will return).
However, I removed the newline, I forgot about that.

Reply:
So I tried with `-pedantic -Wall -Wextra`, which is pretty much every remotely reasonable warning there is, and I couldn't get neither `clang` nor `gcc` to emit a warning about a missing return statement. However, it did reveal a whole bunch of other warnings that we missed during review:

```
install_test.cpp: In function ‘int main(int, char**)’:
install_test.cpp:59:31: warning: static_assert without a message only available with -std=c++17 or -std=gnu++17 [-Wpedantic]
                       type1>::value);
                               ^~~~~
install_test.cpp:73:71: warning: static_assert without a message only available with -std=c++17 or -std=gnu++17 [-Wpedantic]
             std::is_same<gko::Combination<type1>::value_type, type1>::value);
                                                                       ^~~~~
install_test.cpp:80:71: warning: static_assert without a message only available with -std=c++17 or -std=gnu++17 [-Wpedantic]
             std::is_same<gko::Composition<type1>::value_type, type1>::value);
                                                                       ^~~~~
install_test.cpp:86:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::dim<3, type1>{4, 4, 4};
              ^~~~
install_test.cpp:109:54: warning: static_assert without a message only available with -std=c++17 or -std=gnu++17 [-Wpedantic]
         static_assert(gko::is_complex<testType>() == false);
                                                      ^~~~~
install_test.cpp:119:14: warning: unused variable ‘test’ [-Wunused-variable]
         auto test = gko::layout_type::array;
              ^~~~
install_test.cpp:131:14: warning: unused variable ‘test’ [-Wunused-variable]
         auto test = gko::layout_type::array;
              ^~~~
install_test.cpp:136:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::span{12};
              ^~~~
install_test.cpp:142:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::range<gko::accessor::row_major<decltype(testVar), 2>>(
              ^~~~
install_test.cpp:148:70: warning: static_assert without a message only available with -std=c++17 or -std=gnu++17 [-Wpedantic]
         static_assert(std::is_same<gko::xstd::void_t<double>, void>::value);
                                                                      ^~~~~
install_test.cpp:153:24: warning: unused variable ‘test’ [-Wunused-variable]
         gko::size_type test{12};
                        ^~~~
install_test.cpp:158:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::null_deleter<double>{};
              ^~~~
install_test.cpp:163:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::version_info::get().header_version;
              ^~~~
install_test.cpp:173:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
         auto test = gko::log::executor_data{};
              ^~~~
install_test.cpp:301:14: warning: variable ‘stop_status’ set but not used [-Wunused-but-set-variable]
         auto stop_status = gko::stopping_status{};
              ^~~~~~~~~~~
```

While we can overlook the unused variable warning (even though there's really no need to have them - shouldn't make a difference for the final result), the `static_assert` warnings are actually important. If run on a compiler that is C++11 compliant (the minimal requirement for Ginkgo), but not C++17 compliant, the test will fail to compile.

Reply:
I realized that as well when I turned on all warnings. I also included the C++11 requirement in the makefile.

Reply:
I tried to get the warning back myself, but I did not manage to get it back. Maybe I had return statements, but none at the end. Even in Visual Studio I was unable to get it.